src/main/java/org/mule/extensions/jms/internal/lifecycle/ActiveMQArtifactLifecycleListener.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2023 Salesforce, Inc. All rights reserved.
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.extensions.jms.internal.lifecycle;
+
+import static java.util.Optional.empty;
+import static java.util.Optional.of;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+import org.mule.sdk.api.artifact.lifecycle.ArtifactDisposalContext;
+import org.mule.sdk.api.artifact.lifecycle.ArtifactLifecycleListener;
+
+import java.lang.reflect.Field;
+import java.util.Optional;
+import java.util.Timer;
+
+import org.slf4j.Logger;
+
+/**
+ * {@link ArtifactLifecycleListener} for releasing all the known references associated with the ActiveMQ driver that may lead to a
+ * Thread/ClassLoader leak.
+ *
+ * @since 2.0.0
+ */
+public class ActiveMQArtifactLifecycleListener implements ArtifactLifecycleListener {
+
+  private static final String ABSTRACT_INACTIVITY_MONITOR_CLASSNAME = "org.apache.activemq.transport.AbstractInactivityMonitor";
+  private static final Logger LOGGER = getLogger(ActiveMQArtifactLifecycleListener.class);
+
+  @Override
+  public void onArtifactDisposal(ArtifactDisposalContext artifactDisposalContext) {
+    disposeInactivityMonitorTimers(artifactDisposalContext, artifactDisposalContext.getArtifactClassLoader());
+    disposeInactivityMonitorTimers(artifactDisposalContext, artifactDisposalContext.getExtensionClassLoader());
+  }
+
+  private void disposeInactivityMonitorTimers(ArtifactDisposalContext artifactDisposalContext, ClassLoader classLoader) {
+    // See https://github.com/apache/activemq/pull/1119
+    getReadCheckTimerField(artifactDisposalContext, classLoader).ifPresent(this::disposeReadCheckTimer);
+  }
+
+  private Optional<Field> getReadCheckTimerField(ArtifactDisposalContext artifactDisposalContext, ClassLoader classLoader) {
+    try {
+      Class<?> cls = classLoader.loadClass(ABSTRACT_INACTIVITY_MONITOR_CLASSNAME);
+      if (!isArtifactOrExtensionOwnedClass(artifactDisposalContext, cls)) {
+        // Don't do anything if the class does not belong to the artifact being disposed of.
+        return empty();
+      }
+      return of(cls.getDeclaredField("READ_CHECK_TIMER"));
+    } catch (ClassNotFoundException | NoSuchFieldException e) {
+      // If the class or field is not found, there is nothing to dispose
+      return empty();
+    }
+  }
+
+  private void disposeReadCheckTimer(Field readCheckTimerField) {
+    try {
+      // This accessibility override will work as long as the driver is loaded in the unnamed module or as an automatic module
+      readCheckTimerField.setAccessible(true);
```

Comment:
Can we be sure that it will be loaded there? What happens if it does not? Will the leak continue?

Reply:
There are no short/mid term plans for placing the applications/domains and their dependencies in the module-path. See [this](https://docs.google.com/document/d/1rfyHlCEvUD33vjUeFcf8lZDFHB2HZqrJ63FhTB7uHQo/edit#heading=h.keynukwwgqke) and [this](https://docs.google.com/document/d/1rfyHlCEvUD33vjUeFcf8lZDFHB2HZqrJ63FhTB7uHQo/edit#heading=h.axuwsttsqijf).
Even if we were to do so, this will still work with all currently available versions of the ActiveMQ driver, because it does not provide a `module-info.java` specifying what it exports/opens and what it does not. So, by default, all packages from the driver would still be open for reflection.
If, in the future, there was a version of the driver which does not open everything by default, when the customers were to upgrade the driver version to that one, they will start seeing the warning message being logged, with the `IllegalAccessException` as the cause. The leak will indeed happen. They will be able to revert back to the previous version and report the issue. At which point we hope the developers of the driver have already fixed [the issue](https://github.com/apache/activemq/pull/1119) or provided an API for the cleanup.